### ISSUE TITLE
Implement lint-as-you-type

### ISSUE THREAD

**Issue by weirdan** [ikappas/vscode-phpcs#35]
`vscode-phpcs` lints files on save (and load). It's possible to lint unsaved files by feeding file contents to PHPCS on its STDIN. This might provide for more interactive user experience: seeing if you introduced a problem or fixed one without saving the file.

You may borrow some implementation bits from [AtomLinter/linter-phpcs](https://github.com/AtomLinter/linter-phpcs) that implemented this feature.

**Comment by ikappas** [comment]
@weirdan Have a look at the latest development branch and see if this works ;)

**Comment by weirdan** [comment]
It appears to be working. I'm a bit worried though because I don't see any code to throttle linting. Wouldn't it cause phpcs to be spawned for every keystroke? 
I also observed some linting errors appearing for lines that are no longer there. I guess this could be caused linting starting when lines are still there, but completing after the lines had been already removed.

As a sidenote, you can use fixed encoding if you're feeding text to PHPCS's stdin, just like Atom's counterpart does: AtomLinter/linter-phpcs#236 (see issue linked from there for rationale). That'll fix #34 with a single line of code.

**Comment by ikappas** [comment]
@weirdan After doing some tests, I found that change event is throttled at 200ms. See https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/src/main.ts#L534
I have ported the fixed encoding code on the latest development branch. Let me know how this works!

**Comment by weirdan** [comment]
This actually works pretty good for me.

**Comment by oojacoboo** [comment]
When will we see 0.8.0 get pushed out?  I'd like to see this feature as well.  I've currently tweaked my autosave to give me this effect, but it's not preferred.

Also, I might point out, that there are some rules that fix whitespace that cause the cursor to go from an indented state back to the start of the line.  If this was doing it every 200ms, this feature would be useless since you'd have to start typing then tab in every single line.

**Comment by yukulele** [comment]
how to install this version?

**Comment by ikappas** [comment]
@weirdan @oojacoboo @yukulele I've pushed some changes in the development branch. Can you test these changes ?

**Comment by weirdan** [comment]
I got this running, it appears to be working. Is there anything specific you would like to get tested?

**Comment by oojacoboo** [comment]
@weirdan I'm getting `phpcs: Cannot read property 'length' of null`.  Not sure where to find more information on this error?

**Comment by weirdan** [comment]
@ikappas, oh, I had that problem too. Had to cleanup all unversioned/ignored files, then rebuilt the ext & server an then it worked.

**Comment by oojacoboo** [comment]
Looks like the `.gitignore` is incorrect.

`*/node_modules` should be `node_modules`.

**Comment by oojacoboo** [comment]
Same error here though, after that.  `phpcs: Cannot read property 'length' of null`.

**Comment by weirdan** [comment]
Try fresh git clone

**Comment by oojacoboo** [comment]
@ikappas yep.

**Comment by ikappas** [comment]
@oojacoboo win or mac?

**Comment by oojacoboo** [comment]
@ikappas mac.  It'd be nice if the error output actually told you something helpful.  That's a needle/haystack error message output.

**Comment by ikappas** [comment]
@oojacoboo The error suggests that an array is returned somewhere as null

**Comment by oojacoboo** [comment]
@ikappas okay sure.  But that doesn't really help much.  I don't know if this is within analyzed PHP code, the plugin source, the configuration, etc.  It's virtually impossible to track that down, especially since I don't know much about how this plugin works.

It is working properly on `0.7.0`, however.

**Comment by ikappas** [comment]
@oojacoboo can you post your phpcs related settings.json config?

**Comment by weirdan** [comment]
I suspect it's caused by old 'phpcs/server' folder not being cleaned up on build.

**Comment by oojacoboo** [comment]
```json
// Optional. The name or path of the coding standard to use. Defaults to the one set in phpcs global config.
    "phpcs.standard": "app/vendor/rentpost/coding-standard/phpcs.xml",
```

**Comment by ikappas** [comment]
@weirdan If that is the case then manually delete the server folder and run vscode compile task

**Comment by ikappas** [comment]
@oojacoboo did you try last suggestion?

**Comment by oojacoboo** [comment]
@ikappas I don't know what the "server folder" is.  Where is that?

**Comment by ikappas** [comment]
@oojacoboo phpcs/server

**Comment by ikappas** [comment]
I just pushed some major new features in `develop` branch that need testing and include: 

- ability to set phpcs.executablePath
- ability to set phpcs.showWarnings 
- ability to set phpcs.showSources 
- ability to enable phpcs.trace.server with messages or verbose option for tracing errors.
- ability to set phpcs.composerJsonPath to something other that workspaceRoot/composer.json

Your feedback is very welcome!

**Comment by oojacoboo** [comment]
@ikappas `phpcs: ENOENT: no such file or directory, lstat '/Volumes/repos/foo/web/composer.json'`

`composer.json` is located in a sub directory of the project, one level deeper.  Not sure why this should be triggering a `phpcs` issue at any rate.

**Comment by ikappas** [comment]
@oojacoboo Good to hear you got your first error sorted out.

I assume that your workspace root is `/Volumes/repos/foo/web/` so by default the plugin looks for a composer.json file there to automatically resolve the phpcs executable path. It should ignore the missing composer.json and move to searching your global path so I'll have a look.

Meanwhile, since your composer is located one level deeper i.e in a folder named `bar` try to to set the following setting: 
```json
"phpcs.composerJsonPath": "bar/composer.json"
```

**Comment by ikappas** [comment]
@oojacoboo I fixed the issue when composer.json was not found and pushed the code to `develop`

**Comment by oojacoboo** [comment]
@ikappas tested the latest commits.  No issues here.  

`phpcs.showSources` is awesome - much needed to perfect our configs without having to go to the terminal and run `phpcs` separately.

And, of course, linting as you type is huge!

Let's get this thing deployed!

**Comment by JClerc** [comment]
The new features sound awesome! When will this be deployed?

**Comment by ikappas** [comment]
@JClerc Working on multi-root support before releasing 1.0.0

**Comment by ikappas** [comment]
I just pushed multi-root support in develop branch that need testing

**Comment by ikappas** [comment]
I've just pushed 1.0.0-beta.6 on the development branch adding automatic configuration search. This is the last feature for the 1.0.0 release which needs some testing before 1.0.0 launch.

**Comment by ikappas** [comment]
I just released vscode-phpcs 1.0.0 which supports the VSCode 1.20+ and includes many enhancements most notable of which are:

1. multi root support
2. lint as you type support
3. automatic configuration search support
4. ability to set phpcs.executablePath
5. ability to set phpcs.showWarnings
7. ability to set phpcs.showSources
8. ability to set phpcs.composerJsonPath to something other that workspaceRoot/composer.json
9. ability to enable phpcs.trace.server with messages or verbose option for tracing errors.

I would like to give my thanks to all the people that helped track down issues in the previous betas and hope the new features help you get things done easier.

On that note, I am closing this issue to properly track any new issues in the new release. If you find this issue persists or any other issue please feel free to open a new ticket for v.1.0.0.
